docs/units/type_hints.rst

```diff
@@ -0,0 +1,136 @@
+
+::
+
+   >>> import typing as T
+   >>> import astropy.units as u
+   >>> from astropy.units import Quantity
+   >>> from astropy.units._typing import Annotated
+
+
+.. _quantity_type_annotation:
+
+Quantity Type Annotation
+========================
+
+A |Quantity| can easily be used as a type annotation,
+
+::
+
+   >>> x: Quantity = 2 * u.km
+
+or as a function annotation.
+
+::
+
+   >>> def func(x: Quantity) -> Quantity:
+   ...     return x
+
+However, depending on the computer, this frequently throws an exception when compiling Sphinx docs. A safer method to annotate Quantities is by leveraging the :mod:`~typing` library to create a type variable (note the ``~``).
+
+::
+
+   >>> QuantityType = T.TypeVar("QuantityType", bound=Quantity)
+
+Anything annotated with ``QuantityType`` is expected to be a |Quantity| or subclass thereof.
+
+Repeating the above examples,
+
+::
+
+   >>> x: QuantityType = 2 * u.km
+
+or as a function annotation.
+
+::
+
+   >>> def func(x: QuantityType) -> QuantityType:
+   ...     return x
+
+
+Preserving Units
+^^^^^^^^^^^^^^^^
+
+While the above annotations are useful for annotating the **value**'s type, it does not inform us of the other most important attribute of a |Quantity|: the **unit**.
+
+Unit information may be included by the syntax ``Quantity[unit or "physical_type", other_meta]``. **The unit metadata must be first.**
+
+::
+
+   >>> Quantity[u.m]
+   Annotated[~<class 'astropy.units.quantity.Quantity'>, Unit("m")]
+
+   >>> Quantity["length"]
+   Annotated[~<class 'astropy.units.quantity.Quantity'>, 'length']
+
+   >>> Quantity[u.m, "Call", "me", "Ishmael."]
+   Annotated[~<class 'astropy.units.quantity.Quantity'>, Unit("m"), 'Call', 'me', 'Ishmael.']
+
+See :ref:`manual_annotation` for explanation of ``Annotated``
+
+These can also be used on functions
+
+::
+
+   >>> def func(x: Quantity[u.kpc, "input"]) -> Quantity[u.m, "output"]:
+   ...     return x << u.m
+
+
+.. _manual_annotation:
+
+Manually Constructing an Annotation
+===================================
+
+The ``Quantity[unit or "physical_type", other_meta]`` annotations can also be manually constructed. While this is not recommended, we show it here to explain the conventions and how it relates to the python :mod:`typing` library.
+
+Unit-aware |Quantity| annotations are an annotation to the type using the ``Annotated`` class.
+
+In Python 3.9+ ``Annotated`` is built into
+:mod:`~typing`. For older python versions, the optional dependency
+``typing_extensions`` provides the same class. When neither are
+available we default to a run-time only compatibility class (in
+``astropy.units._typing``) with reduced functionality. The best
+implementation is automatically detected and used when importing ``Annotated`` from ``astropy.units._typing``.
+
+::
+
+  >>> Annotated[QuantityType, u.m, "I am a leaf on the wind"]
+  Annotated[~QuantityType, Unit("m"), 'I am a leaf on the wind']
+
+**The unit metadata must be first.**
```

Comment:
No need for this sentence given the one that follows.

Reply:
👍

Reply:
I've been thinking about this and I think it's a good implementation for this PR. We should more closely examine the syntax for numpy ``ndarray`` because it would be nice to figure out a compatible way of easily doing multiple annotations. One trivial solution is ``Quantity[(unit1, unit2, physicaltype, ...)]`` . It's not quite as clean looking as ``Quantity[unit1, unit2, physicaltype, ...]``, nor does it directly match the parameter annotation syntax which eschews the parentheses.
However the tuple option makes future compatibility with numpy dtype stuff and supporting arbitrary extensions by allowing non-unit metadata quite easy. The non-tuple option would be more challenging.